vss-cats/src/main/scala/com/virtuslab/vss/cats/stats/services/Stats.scala

```diff
@@ -0,0 +1,30 @@
+package com.virtuslab.vss.cats.stats.services
+
+import upickle.default.*
+import cats.implicits.*
+import com.virtuslab.vss.common.Event
+import fs2.kafka.KafkaProducer
+import upickle.default.*
+import cats.Monad
+import cats.effect.std.AtomicCell
+
+/**
+ * Service responsible for handling events.
+ */
+sealed abstract class Stats[F[_]] {
+  def getLatestEvents(amount: Int): F[List[Event]]
+  def addEvent(eventStr: String): F[Unit]
+}
+
+object Stats {
```

Comment:
> It's obviously not 70%, even in this service that has almost no logic.

In this particular case it's about 70%:

Abstract class + object ~13 lines

```
sealed abstract class Stats[F[_]] {
  def getLatestEvents(amount: Int): F[List[Event]]
  def addEvent(eventStr: String): F[Unit]
}

object Stats {
  def make[F[_]: Monad](
    eventsStore: AtomicCell[F, Seq[Event]]
  ): Stats[F] =
    new Stats[F] {
      def getLatestEvents(amount: Int): F[List[Event]] = eventsStore.get.map(_.take(amount).toList)

      def addEvent(eventStr: String): F[Unit] = eventsStore.update(read[Event](eventStr) +: _)
    }
}
```

Single class - 4 lines:

```
class Stats(eventsStore: AtomicCell[IO, Seq[Event]]) {
  def getLatestEvents(amount: Int): IO[List[Event]] = eventsStore.get.map(_.take(amount).toList)
  def addEvent(eventStr: String): IO[Unit] = eventsStore.update(read[Event](eventStr) +: _)
}
```

---

> It you read the base part of the code closely, you will find that it's not IO. It's actually [A] =>> Kleisli[IO, Span[IO], A].

Hmm, could you point it out? I see IO 🤔 

`services = Services.make[IO](res.eventsStore)`

I've made a test and got rid of generic param and it works. You can find it [here](https://github.com/VirtusLab/vss/tree/cats-io-instead-of-generic).

> Specifically, it doesn't allow for switching the implementation monad for free.

It's true. I'm wondering if it's worth here - generic code always generates cost and I understand the goal of this project is to make using Scala easier. My impression is generic here makes the code seems to be more complex than it really is.

---

But it's a matter of taste so if you reckon that current solution is better I won't insist.

Reply:
U probably misread my comment:
> It you read the **BASE** part of the code closely, you will find that it's not `IO`. It's actually `[A] =>> Kleisli[IO, Span[IO], A]`.

What I meant about the percentage is that it is clearly a highly simplified implementation, in a normal case it will obviously be an even more insignificant number.

> In this particular case it's about 70%:

I clearly see 30 lines vs 22 lines. (I removed an unused import) Not even close to 50%.

Reply:
> It's not. Hardcoding IO is not the way to write code in cats-effect. If you think that u can make the code easier hardcoding IO like that, then try doing it in BASE instead of STATS.

Yes, you're right 👍 I didn't got what you mean here:

> It you read the base part of the code closely

---

> I clearly see 30 lines vs 22 lines. (I removed an unused import) Not even close to 50%.

Please look at my comment above. I didn't count imports as they're not crucial to understand the code.

> Abstract class + object ~13 lines

> Single class - 4 lines